qiskit/pulse/transforms/frames.py

```diff
@@ -0,0 +1,140 @@
+from qiskit.pulse.transforms.canonicalization import block_to_schedule
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse import channels as chans, instructions
+from qiskit.pulse.frame import FramesConfiguration
+
+
+def resolve_frames(
+    schedule: Union[Schedule, ScheduleBlock], frames_config: FramesConfiguration
+) -> Schedule:
+    """
+    Parse the schedule and replace instructions on Frames by instructions on the
+    appropriate channels.
+
+    Args:
+        schedule: The schedule for which to replace frames with the appropriate
+            channels.
+        frames_config: An instance of FramesConfiguration defining the frames.
+
+    Returns:
+        new_schedule: A new schedule where frames have been replaced with
+            their corresponding Drive, Control, and/or Measure channels.
+
+    Raises:
+        PulseError: if a frame is not configured.
+    """
+    if frames_config is None or not frames_config:
+        return schedule
+
+    if isinstance(schedule, ScheduleBlock):
+        schedule = block_to_schedule(schedule)
+
+    resolved_frames = {}
+    sample_duration = None
+    for frame, frame_def in frames_config.items():
+        resolved_frames[frame] = ResolvedFrame(frame, frame_def)
+
+        # Extract shift and set frame operations from the schedule.
+        resolved_frames[frame].set_frame_instructions(schedule)
+        sample_duration = frame_def.sample_duration
+
+    if sample_duration is None:
+        raise PulseError("Frame configuration does not have a sample duration.")
+
+    # Used to keep track of the frequency and phase of the channels
+    channel_trackers = {}
+    for ch in schedule.channels:
+        if isinstance(ch, chans.PulseChannel):
+            channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    sched = Schedule(name=schedule.name, metadata=schedule.metadata)
+
+    for time, inst in schedule.instructions:
+        if isinstance(inst, instructions.Play):
+            chan = inst.channel
+
+            if inst.signal.frame is not None:
+                frame = inst.signal.frame
+
+                if frame not in resolved_frames:
+                    raise PulseError(f"{frame} is not configured and cannot " f"be resolved.")
+
+                resolved_frame = resolved_frames[frame]
+
+                frame_freq = resolved_frame.frequency(time)
+                frame_phase = resolved_frame.phase(time)
+
+                # If the frequency and phase of the channel has already been set once in
+                # The past we compute shifts.
+                if channel_trackers[chan].is_initialized():
+                    freq_diff = frame_freq - channel_trackers[chan].frequency(time)
+                    phase_diff = frame_phase - channel_trackers[chan].phase(time)
+
+                    if abs(freq_diff) > resolved_frame.tolerance:
+                        shift_freq = instructions.ShiftFrequency(freq_diff, chan)
```

Comment:
This is nitpick but why only shift_freq is defined (why not shift_phase below)?

Reply:
This seems outdated after the previous commits.